### PR TITLE
Report all-empty top level statements.

### LINKED ISSUE #53472: Stray semicolon above namespace keyword causes warning CS7022 and unexpected exit
### Description
Expected behavior:
Either compiler error to be thrown or for everything to run as expected without a sudden exit.

Actual behavior:
Having a stray semicolon above the namespace keyword causes CS7022. A reproducable example is below:
```cs
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
; // This will cause CS7022

namespace Tomoe.Commands.Public {
```
This can be placed in any C# file, and the runtime warning will occur, causing the program to suddenly exit with code 0. Opening up the debugger, it seems it exits right after `System.Runtime.dll` loads. It seems several people over on this [StackOverflow Answer](https://stackoverflow.com/a/65679104/10942966) have also been having this issue.

### Configuration
.NET version: 5.0.202
OS: Windows 10 and Ubuntu 20.04
Arch: x64
Specific to configuration: Untested.

### Regression?
Haven't thoroughly tested. On .netcore3.1 in Visual Studio, a syntax error is thrown with the following message: `Feature "top-level statements" is not available in C# 8. Please use language version 9 or higher`

### Other information
I start the program, I get the warning, then it suddenly exits with code 0. Yes, the same behaviour happens when I'm in Gnome Terminal (Ubuntu) or Windows Terminal/Powershell (Windows 10)
image
image

### Simple Solution
Remove stray semicolons. No stack traces will be thrown. A great way to teach fellow peers to always review PR's instead of accepting them blindly.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #53472.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs [hunk 1]
```diff
@@ -74,6 +75,11 @@ private DeclarationTreeBuilder(SyntaxTree syntaxTree, string scriptClassName, bo
                     firstGlobalStatement ??= global;
                     var topLevelStatement = global.Statement;
 
+                    if (!topLevelStatement.IsKind(SyntaxKind.EmptyStatement))
+                    {
+                        hasNonEmptyGlobalSatement = true;
+                    }
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs:81]: Should the error be reported if the top-level program contains only local functions?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs:81]: > Should the error be reported if the top-level program contains only local functions?

[https://github.com/dotnet/roslyn/issues/53472#issuecomment-866502606](https://github.com/dotnet/roslyn/issues/53472#issuecomment-866502606) - "LDM decided to make it an error if all top-level statements are empty." I do not see LDM deciding anyjthing about local function statements.

Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs:81]: @AlekseyTs Yeah I see the PR matches LDM decision, but I mean, was that considered?

The goal of the error is to tell the user "your program will execute nothing and just return 0 immediately", which is the case when the top-level statements are local functions.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs:81]: > The goal of the error is to tell the user "your program will execute nothing and just return 0 immediately", 

I do not think that is the goal, an entry point function is always executed. There is plenty of ways to write code that does nothing even when it has non-empty statements. We were addressing specific customers scenario and I don't think there was a goal to generalize the issue. If you feel that the issue should be generailized, open an issue and provide your reasons. I beleive a local function that is never called produces a warning, in my opinion that is goog enough to give a hint that the code probably doesn't do what was intended.

**Review by cston — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@jcouv, @dotnet/roslyn-compiler For the second review.

File: src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs [hunk 2]
```diff
@@ -98,7 +104,17 @@ private DeclarationTreeBuilder(SyntaxTree syntaxTree, string scriptClassName, bo
...
+                var diagnostics = ImmutableArray<Diagnostic>.Empty;
+
+                if (!hasNonEmptyGlobalSatement)
+                {
+                    var bag = DiagnosticBag.GetInstance();
+                    bag.Add(ErrorCode.ERR_SimpleProgramIsEmpty, ((EmptyStatementSyntax)firstGlobalStatement.Statement).SemicolonToken.GetLocation());
+                    diagnostics = bag.ToReadOnlyAndFree();
+                }
+
+                childrenBuilder.Add(CreateSimpleProgram(firstGlobalStatement, hasAwaitExpressions, isIterator, hasReturnWithExpression, diagnostics));
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs:116]: nit: extra empty line below

File: src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs [hunk 3]
```diff
@@ -8819,5 +8819,73 @@ void M()
...
+        {
+            var text = @"
+;
+System.Console.WriteLine(""Hi!"");
+;
+";
+
+            var comp = CreateCompilation(text, options: TestOptions.DebugExe, parseOptions: DefaultParseOptions);
+            CompileAndVerify(comp, expectedOutput: "Hi!");
+        }
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs:8889]: nit: Consider adding a test from the original issue `using System; ;` or something like it

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 2)

**Pull request merged**